From 7f5b98999e6fd7f50902cfbfaca60de5cfbc5b9a Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Thu, 4 Jul 2019 22:12:12 -0700
Subject: [PATCH 0643/1851] v4l: xilinx: sdirxss: Uninitialize the subsystem
 correctly

commit e7f3a7a70395a423d7bb0c81507576703a96a726 from
https://github.com/Xilinx/linux-xlnx.git

When the probe fails or driver is removed, uninitialize the subsystem
correctly by disabling the global interrupt, individual interrupts, core
and bridges in the probe() and remove() functions.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-sdirxss.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index 07dfaa94001a..4af98631a1e9 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -1999,6 +1999,8 @@ static int xsdirxss_probe(struct platform_device *pdev)
 error:
 	v4l2_ctrl_handler_free(&xsdirxss->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
+	xsdirx_globalintr(core, false);
+	xsdirx_disableintr(core, XSDIRX_INTR_ALL_MASK);
 clk_err:
 	clk_bulk_disable_unprepare(core->num_clks, core->clks);
 	return ret;
@@ -2013,6 +2015,12 @@ static int xsdirxss_remove(struct platform_device *pdev)
 	v4l2_async_unregister_subdev(subdev);
 	v4l2_ctrl_handler_free(&xsdirxss->ctrl_handler);
 	media_entity_cleanup(&subdev->entity);
+
+	xsdirx_globalintr(core, false);
+	xsdirx_disableintr(core, XSDIRX_INTR_ALL_MASK);
+	xsdirx_core_disable(core);
+	xsdirx_streamflow_control(core, false);
+
 	clk_bulk_disable_unprepare(core->num_clks, core->clks);
 
 	return 0;
-- 
2.31.1

